src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -11806,6 +11806,102 @@ struct BroadcastInDimIsReshape final
   }
 };
 
+struct PadConcatToConcatPad
+    : public OpRewritePattern<stablehlo::ConcatenateOp> {
+  using OpRewritePattern<stablehlo::ConcatenateOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(stablehlo::ConcatenateOp concatOp,
+                                PatternRewriter &rewriter) const override {
+
+    // Check if all operands are pad ops with the same padding value
+    SmallVector<stablehlo::PadOp> padOps;
+    Value padValue;
+
+    for (Value operand : concatOp.getOperands()) {
+      auto padOp = operand.getDefiningOp<stablehlo::PadOp>();
+      if (!padOp)
+        return failure();
+
+      if (padOps.empty()) {
+        padValue = padOp.getPaddingValue();
+      } else if (padValue != padOp.getPaddingValue()) {
+        return failure(); // Different padding values not supported
+      }
+
+      padOps.push_back(padOp);
+    }
+
+    int64_t concatDim = concatOp.getDimension();
+    int64_t rank = padOps[0].getEdgePaddingLow().size();
+
+    // Compute smallest common padding for all tensors
+    SmallVector<int64_t> commonLowPadding(rank,
+                                          std::numeric_limits<int64_t>::max());
+    SmallVector<int64_t> commonHighPadding(rank,
+                                           std::numeric_limits<int64_t>::max());
+    SmallVector<int64_t> interiorPadding(rank, 0);
+
+    // Find minimum padding across all inputs (conservative common padding)
+    for (auto padOp : padOps) {
+      for (int64_t dim = 0; dim < rank; ++dim) {
+        commonLowPadding[dim] =
+            std::min(commonLowPadding[dim], padOp.getEdgePaddingLow()[dim]);
+        commonHighPadding[dim] =
+            std::min(commonHighPadding[dim], padOp.getEdgePaddingHigh()[dim]);
+      }
+    }
+
```

Comment:
you need to return early here if all of the common paddings are zero

Reply:
I'll add this